### PR TITLE
Add section checking based on sig type and `auth` section for auth sigs

### LINKED ISSUE #22: Check that signatures have the right stanzas in their bodies based on sig types.
This should be part of `_v_check`.  Codify the possibilities in the library.

CC: @lord @oconnor663

### PR DESCRIPTION
**Pull request by lord**
mind taking a look? @maxtaco @oconnor663 

Fixes #22

This makes it so that, for instance, a signature with the type `revoke` can't contain a `sibkey` section. I've tested this with ~1000 sigchains (for a total of ~7000 signatures) downloaded from the live site. It passes all of them, and I'd be glad to test against more if we want to double check. The Keybase server tests also pass with these changes installed.

Some questions/possible todos:

- [x] Is `stanza` the correct term, or would `section` be better?
- [x] Will `sibkey` or other types of signatures need to have `revoke` statements in the future? Right now this PR only allows `revoke` statements in `revoke`, `cryptocurrency`, and `web_service_binding`.
- [x] There are some signature types (like `dualkey` and `auth`) in there that I'm not familiar with. Although we're passing the server tests, I'm not 100% sure I defined the stanzas correctly for those, so a double-check would be great.
- [x] This just checks the top-level stanzas in `body` to make sure they're right, not any deeper. Is that the correct behavior?
- [x] Still needs to be signed/versioned/changelogged

### CODE REVIEW

File: src/base.iced [hunk 1]
```diff
@@ -197,17 +197,40 @@ class Base
...
       err = @_v_check_key key
     cb err
 
   #------
 
+  _required_stanzas : () -> ["key", "type", "version"]
+
+  #------
+
+  _optional_stanzas : () -> ["client", "merkle_root"]
```
Comment (oconnor663) [inline src/base.iced]: Are these really optional going forward, or just optional in the past? Can we make them required now?

File: src/announcement.iced [hunk 2]
```diff
@@ -12,7 +12,9 @@ exports.Announcement = class Announcement extends Base
 
   _type : () -> constants.sig_types.announcement
 
-  _json : () -> 
+  _required_stanzas : () -> super.concat(["announcement"])
```
Comment (oconnor663) [inline src/announcement.iced]: I was surprised to learn that `super` is implicitly the same as `super()`. Could we use the more explicit version?

File: src/base.iced (hunk 1 above)
Comment (maxtaco) [inline src/base.iced]: we'd have to change: (1) the web site; (2) old versions of node client; and (3) the "hardcore" mode of the Web site.

**Comment by oconnor663** [comment]
- "Stanza" sounds fine to me. 
- I don't think any of the key delegation links need to also be a revoke.
- Just checking the first level sounds fine to me too.

File: src/base.iced (hunk 1 above)
Comment (oconnor663) [inline src/base.iced]: Ah ok, I was misinterpreting this as a check that got performed only on *new* links.

**Comment by maxtaco** [comment]
1. I like `section` better, you're right.
1. Yes, sibkey and subkey can have `revoke` statements.
1. `dualkey` is on the chopping block, it will be gone soon.
1. `auth` is for public key auth. has one required section and no optional.

File: src/announcement.iced (hunk 2 above)
Comment (lord) [inline src/announcement.iced]: Ah, yea, I think it's a "try to be more like Ruby" thing, `super` implicitly adds all the arguments. Agreed that explicit is better though, should be updated!

**Comment by lord** [comment]
Okay, I've renamed `stanza` to `section`, added `revoke` sections to `sibkey` and `subkey`, and added parens.

**Comment by oconnor663** [comment]
Relaying from @maxtaco: Looks good to merge. There's something about how auth signatures work that Max wants to change, but he'll do that after this is in.

**Comment by lord** [comment]
@oconnor663: :+1: cool! still needs a CHANGELOG and SIGNED, should those be in the PR or once it's in `master`?

**Comment by maxtaco** [comment]
fold this in with the fixes to auth before making a new release...

**Comment by maxtaco** [comment]
LGTM :+1: ship it and i will cut a new version

**Pull request merged**